File: 82.js
class PriorityQueue {
    constructor() {
        this.data = [];
    }

    push(item) {
        this.data.push(item);
        this.data.sort((a, b) => {
            if (a.priority === b.priority) {
                if (a.time === b.time) return a.id.localeCompare(b.id);
                return a.time - b.time;
            }
            return b.priority - a.priority;
        });
    }

    pop() {
        return this.data.shift();
    }

    top() {
        return this.data[0];
    }

    size() {
        return this.data.length;
    }
}

const fs = require("fs");
const input = fs.readFileSync(0, "utf-8").trim().split("\n");

let Q = Number(input[0]);
let pq = new PriorityQueue();

let idx = 1;
while (Q--) {
    let parts = input[idx++].split(" ");
    let cmd = parts[0];

    if (cmd === "ADD") {
        pq.push({
            id: parts[1],
            time: Number(parts[2]),
            priority: Number(parts[3])
        });
    }
    else if (cmd === "RUN") {
        if (pq.size() === 0) console.log("EMPTY");
        else console.log(pq.pop().id);
    }
    else if (cmd === "TOP") {
        if (pq.size() === 0) console.log("EMPTY");
        else console.log(pq.top().id);
    }
    else if (cmd === "COUNT") {
        console.log(pq.size());
    }
}

File: 82.py
import heapq

pq = []

Q = int(input())
for _ in range(Q):
    parts = input().split()
    cmd = parts[0]

    if cmd == "ADD":
        job_id = parts[1]
        t = int(parts[2])
        p = int(parts[3])
        heapq.heappush(pq, (-p, t, job_id))

    elif cmd == "RUN":
        if not pq:
            print("EMPTY")
        else:
            _, _, job_id = heapq.heappop(pq)
            print(job_id)

    elif cmd == "TOP":
        if not pq:
            print("EMPTY")
        else:
            print(pq[0][2])

    elif cmd == "COUNT":
        print(len(pq))
